Make the profile_image api more restful.

jcdyer opened the pull request:
https://openedx.atlassian.net/browse/MA-1280

Deprecate the views at `/api/profile_images/v1/{username}/upload` and
`/api/profile_images/v1/{username}/remove` and replace them with
GET and POST methods at `/api/user/v1/accounts/{username}/image`

@nasthagiri

jcdyer commented:
I left the bulk of the code in /openedx/core/djangoapps/profile_image, even though that only contains deprecated views now.  If there's a strong desire for it, I could move all the profile_image code into user_api.  I didn't want to make too invasive of a change for no good reason, though, so I left most things where they are.

jcdyer commented:
@catong I documented the old views as deprecated, but I'm not sure if there's an established convention I should be using for that.  Please take a look at the doc changes and offer critiques.

jcdyer commented:
@jimabramson Please review

nasthagiri commented on openedx/core/djangoapps/profile_images/tests/test_views.py:
```diff
@@ -160,22 +162,22 @@ def check_upload_event_emitted(self, old=None, new=TEST_UPLOAD_DT):
     def test_anonymous_access(self, mock_log):
         """
-        Test that an anonymous client (not logged in) cannot POST.
+        Test that an anonymous client (not logged in) cannot call GET or POST.
```
Don't forget `DELETE`!

nasthagiri commented on openedx/core/djangoapps/profile_images/tests/test_views.py:
```diff
@@ -405,13 +385,79 @@ def test_remove_staff(self, mock_log):
         self.assert_no_events_were_emitted()
+
+
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Profile Image API is only supported in LMS')
```
Curious: If this API is only valid in LMS, why is it located in the `openedx` folder?

cahrens replied:
Good question. Around the time that we wrote this API, there was a push to put new functionality in openedx instead of common (to make them actual django apps and get away from the oddness of common). We did envision that the APIs could be useful in CMS, but since we had no current use case for it, we chose not to add the URLs to the CMS until a use case presented itself.

@andy-armstrong does that match your memory as well?

andy-armstrong replied:
Yes, that's correct. There is a longer term plan to add profile images to Studio, so that course teams can use them. In fact, @marcotuts has a Hackathon proposal to do exactly this:

https://openedx.atlassian.net/wiki/display/OPEN/Hackathon+11%3A+Return+of+the+Con+Hackathon

nasthagiri replied:
Gotcha.  Thanks.

nasthagiri commented on openedx/core/djangoapps/profile_images/tests/test_views.py:
```diff
@@ -405,13 +385,79 @@ def test_remove_staff(self, mock_log):
+
+    def setUp(self):
+        super(DeprecatedProfileImageRemoveTestCase, self).setUp()
+        with make_image_file() as image_file:
```
Is this `setUp` method needed?  Neither of the 2 test methods below require this setUp, right?

nasthagiri commented on openedx/core/djangoapps/profile_images/tests/test_views.py:
```diff
@@ -405,13 +385,79 @@ def test_remove_staff(self, mock_log):
+
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Profile Image API is only supported in LMS')
+@mock.patch('openedx.core.djangoapps.profile_images.views.log')
+class DeprecatedProfileImageRemoveTestCase(ProfileImageEndpointTestCase):
```
Consider creating a common test Mixin class for `DeprecatedProfileImageUploadTestCase` and `DeprecatedProfileImageRemoveTestCase` to reduce code duplication.

nasthagiri commented on openedx/core/djangoapps/user_api/urls.py:
```diff
@@ -12,18 +12,23 @@
         name="preferences_detail_api"
     ),
-)
+)
```
Needs EOL.

nasthagiri commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
-          upload his or her own profile image.
+        The requesting user must be signed in.  Users can only add profile
+        images to their own account.  Users with staff access can remove
+        profile images for other user accounts.  All other user can
```
s/user/users

nasthagiri commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
 
-        * If no user matches the "username" parameter, the request returns an HTTP
-          404 "Not Found" response.
+        When the requesting user tries to upload the image for a different
```
A success response is also possible `when the requesting user tries to upload the image for a different user`.

nasthagiri commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
-        * If no user matches the "username" parameter, the request returns an HTTP
-          404 "Not Found" response.
+        When the requesting user tries to upload the image for a different
+        user, the request may return any of the following:
```
A success response is also possible `when the requesting user tries to upload the image for a different user`.  Perhaps just insert an `also` in `may return`.

jcdyer replied:
I don't think so.  The previously existing documentation on the upload view says that a regular user will receive a 404 when trying to post to any username other than themself, and a staff user will get a 404 for a non-existent username and a 403 for an existing user.  I think the idea is that a staff user can delete an inappropriate profile_image, but can't add a particular image for another user.  The test ProfileImageViewPostTestCase.test_upload_staff (L226) confirms this; it demonstrates that a staff user gets a 403 when trying to upload an image for another user.

nasthagiri commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+        * If no user matches the "username" parameter, the request returns an
+          HTTP 404 "Not Found" response.
+
+        * If a user matches the "username" parameter and the user has staff
```
Nit: Although outside the scope of this PR, it would be nice to keep the permission details outside of the API documentation.  Specifically, instead of saying `the user has staff access` we should say `the user has delete access to the other user's profile image`.

nasthagiri commented:
@jcdyer Looks good overall.  I have completed my first pass.  Please note you have a few quality issues to address as well.  Thanks.

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+    **Use Cases**
 
-        * Upload an image for the user's profile.
+        Add or remove profile images associated with the user's account.
```
"Add or remove profile images associated with user accounts."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+        The requesting user must be signed in.  Users can only add profile
+        images to their own account.  Users with staff access can remove
+        profile images for other user accounts.  All other users can
+        only remove their own profile images.
```
Transpose "only" and "remove": "All other users can remove only their own profile images."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
 
-        When the requesting user tries to upload the image for a different user, the
-        request returns one of the following responses.
+    **Example POST Responses**
```
Can we add back an intro? "When the requesting user attempts to upload an image, the request returns one of the following responses."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
-        * If the requesting user has staff access, the request returns an HTTP 403
-          "Forbidden" response.
+        * If the upload could not be performed, the request returns an HTTP 400
+          "Bad Request" response with more information.
```
What is the additional information that's included?

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
 
-        * If no user matches the "username" parameter, the request returns an HTTP
-          404 "Not Found" response.
+        When the requesting user tries to upload the image for a different
```
"When" -> "If" : "If the requesting user..."
"the image" -> "an image"

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
-        * If no user matches the "username" parameter, the request returns an HTTP
-          404 "Not Found" response.
+        When the requesting user tries to upload the image for a different
+        user, the request may also return any of the following:
```
"the request might also return one of the following responses."

jcdyer replied:
Actually it's not a "might also" in this case, because the 204 and 400 responses never occur in these cases only 404s and (for staff users) 403s.  How about prefacing the first two bullets with: 

> When the requesting user attempts to upload an image for their own account, the request returns one of the following responses:

and prefacing the last three bullets with:

> If the requesting user tries to upload an image for a different user, the request returns one of the following responses:

catong replied:
Sounds great. Thanks, @jcdyer!

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
 
-        * If the upload is successful, the request returns an HTTP 204 "No Content"
-          response with no additional content.
+        * If the user exists and the requesting user does not have staff
```
"and" -> "but" and clarify the first instance of "user": 
"If the user whose profile image is being uploaded exists, *but* the requesting user does not have staff access..."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+
+        * If the user exists and the requesting user has staff access, the
+          request returns an HTTP 403 "Forbidden" response. ??? Seems this is possible...
+
```
Still to be confirmed?

jcdyer replied:
Oops.  Thanks.  I've confirmed the documentation was accurate.  Will update.

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+        * If the user exists and the requesting user has staff access, the
+          request returns an HTTP 403 "Forbidden" response. ??? Seems this is possible...
+
+    **Example DELETE Responses**
```
Can we add an intro here too please: ""When the requesting user attempts to remove an image, the request returns one of the following responses."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+    **Example DELETE Responses**
+
+        * If the image could not be removed, the request returns an HTTP 400
+          "Bad Request" response with more information.
```
What is the additional info?

jcdyer replied:
@catong How specific do you want me to be.  I can update it to "an HTTP 400 'Bad Request' response with information about why the request failed," which would be pretty easy.  If you'd like to see specifics about what kind of errors we would expect to see, I'd need to dig deeper into the code base.

catong replied:
"information why the request failed" is sufficient. It's much less mysterious than "more information" :-)

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+        different user, the view may also return one of the following
+        responses:
+
+        * If the user does not have staff access, the request returns an HTTP
```
"If the requesting user does not have staff access..."

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+          HTTP 404 "Not Found" response.
+
+        * If a user matches the "username" parameter and the user has staff
+          access, the requested user is deleted, and the request returns an
```
" the requested user is deleted" -> "...the requested profile image is deleted"?

jcdyer replied:
Right.  I'm also thinking "requested user" -> "specified user" would be a little clearer (more visual contrast than requested/requesting).

catong replied:
Yes, sounds great, thanks!

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -35,49 +35,75 @@ def _make_upload_dt():
+
+        * If a user matches the "username" parameter and the user has staff
+          access, the requested user is deleted, and the request returns an
+          HTTP 204 "No Content" response with no additional content
```
Missing period at end of sentence

catong commented on openedx/core/djangoapps/profile_images/views.py:
```diff
@@ -121,9 +147,87 @@ def post(self, request, username):
+
+        This view is deprecated.  Use POST
+        /api/user/v1/accounts/{username}/image instead.
+
```
For deprecated views, I think we should remove the use case, example request, example responses text, leaving just the warning and the "Use XX instead" text that you have in lines 178-179. Do you agree @jcdyer ?

jcdyer replied:
Would it be helpful to keep the example request around to show which URL routes are being deprecated?  We could make it part of the warning section, but I think it will be useful for API users to have access to a little more information about what is going away.  (DRF might provide this in the autogenerated parts of the docs.  I'm not sure.) @aleffert  or @mikekatz might have an opinion on this, too.

jcdyer commented:
@mikekatz @aleffert Please take a look at this PR.

mikekatz commented:
looks good from my end. let me know when it's mergeed

jimabramson commented on openedx/core/djangoapps/user_api/urls.py:
```diff
@@ -12,17 +12,22 @@
     '',
     url(
-        r'^v1/accounts/' + USERNAME_PATTERN + '$',
+        r'^v1/accounts/%s$' % (USERNAME_PATTERN,),
```
i believe our convention these days is to use [`str.format`](https://docs.python.org/2/library/stdtypes.html#str.format) rather than interpolation.  (applies to the other entries below as well)

jcdyer replied:
Preference for `r'^v1/accounts/{}$'.format(USERNAME_PATTERN)` vs. `r'^v1/accounts/{username}$'.format(username=USERNAME_PATTERN)` ?

jcdyer replied:
I'm leaning toward `{}`, because it's formatted immediately, and the name of the variable is clear enough, but if there's a convention to name all format args, I'm happy to follow it.

jimabramson commented on openedx/core/djangoapps/profile_images/urls.py:
```diff
@@ -1,5 +1,10 @@
 """
 Defines the URL routes for this app.
+
+NOTE: These views are deprecated.  These routes are superceded by
```
s/supercede/supersede/ (i hope this is not flamebait.)

jcdyer replied:
Dang.  I totally would have lost the spelling bee on that one.

jimabramson commented:
two nitpicks, otherwise :+1: from me.

nasthagiri commented on openedx/core/djangoapps/user_api/urls.py:
```diff
@@ -4,7 +4,7 @@
 from .preferences.views import PreferencesView, PreferencesDetailView
-
+from ..profile_images.views import ProfileImageView
 from django.conf.urls import patterns, url
```
Nit: I believe somewhere in our conventions it says the import order should have the local imports on the bottom and external imports on top.

jcdyer replied:
I saw that, but didn't realize that ..imports were considered non-local.  Should dotted-imports be sorted by descending number of leading dots?

jcdyer replied:
Oh.  You were talking about django.conf.urls.  Got it.

jimabramson replied:
group imports:
* standard library
* external packages (incl third party)
* current package.

alphabetize based on the first non-keyword (e.g. whatever occurs after `import` or `from`)

in other repos we are avoiding relative imports.

nasthagiri commented on openedx/core/djangoapps/profile_images/tests/test_views.py:
```diff
@@ -330,34 +335,12 @@ def check_remove_event_emitted(self):
-        self.assertFalse(mock_log.info.called)
-        self.assert_no_events_were_emitted()
-
-    def test_anonymous_access(self, mock_log):
```
Is the anonymous_access test case no longer verified for the `Delete` method?

jcdyer replied:
I put the squashed the tests together in the Upload test case, but it does test all the methods (POST, DELETE, and even GET, even though there's no actual GET method).

nasthagiri replied:
Nit:
I consider testing the absence of the unsupported_methods as a test for the View.  So, yes, I can see the rationale for testing that only once.
However, I consider verification of a REST method's access as a method-specific test and should be tested for all methods (POST and DELETE).

nasthagiri commented:
@jcdyer Also, please include a link to the JIRA ticket in the description of this PR and the number of the JIRA ticket in the commit message.  We have automatic linking (in JIRA and in our release process) that gets triggered when we do so.

jcdyer commented:
@nasthagiri: I addressed the issues, but didn't refactor the tests, because they were calling different methods, so I would have had to extract out something like `self.call_method = self.client.{post,delete}` and it started to seem uglier than it was helpful.  I'll squash in the morning, if it all looks good to you.

nasthagiri commented:
@jcdyer Completed my 2nd pass.  Looks good.  I'll wait for the squash and the Jenkins green before thumbing.

jcdyer commented:
jenkins run lettuce

nasthagiri commented:
:+1:

nasthagiri commented:
:+1:

nasthagiri commented:
@wajeeha-khalid FYI.  Just tagging you so you are aware of this change being merged in.

jcdyer commented:
jenkins run bokchoy

nasthagiri commented:
CONGRATS!

cahrens commented:
@nasthagiri and @jcdyer I forgot to mention that the profile image load tests need to be updated for this change. They live in load-tests\user_api\profile_images.py.
